### PR TITLE
00995 d regression history

### LINKED ISSUE #995: Implement summary history support for CircleCI Regression
## Summary

We need to support proper reporting of summary history on nightly regression runs in CircleCI. We already have a working implementation which is documented below in this ticket chain.

Please note that the docker image used for CircleCI needs to have the Google Cloud SDK installed and available. 

### CircleCI Example Job Definition

```
######################################################################################################################
  # Job Name:                     jrs_regression
  # Job Version:                  1.0
  # Target Operating Systems:     Ubuntu 18.04 (bionic), Ubuntu 20.04 (focal), CentOS 7, CentOS 8
  #
  # Parameters:
  #       runtime                 a CircleCI executor reference used to control the OpenJDK and other tool versions
  #
  #
  # Description:
  #
  #   Description goes here
  #
  #
  #
  # Executor Requirements:
  #   - git (>= 2.17.1)
  #   - wget (>= 1.19.4)
  #   - zip (>= 3.0)
  #   - unzip (>= 3.0)
  #   - gzip (>= 1.6)
  #   - tar (>= 1.29)
  #   - haveged (>= 1.9.1)
  #   - maven (>= 3.6.1)
  #   - openjdk (>= 12.0.2)
  #   - postgresql-server (>= 10.9)
  #   - gcloud-sdk (>= 323.0.0)
  #
  # References:
  #   - OpenJDK:          https://jdk.java.net/
  #   - Maven:            https://maven.apache.org/
  #
  ######################################################################################################################
  jrs_regression:
    parameters:
      runtime:
        type: executor
      config_file:
        type: string
      regression_path:
        type: string
        default: "regression"
      result_path:
        type: string
        default: "regression/results"
      summary_path:
        type: string
        default: "regression/summaryHistory"
      summary_bucket:
        type: string
        default: "swirlds-circleci-jrs-history"
      result_bucket:
        type: string
        default: "swirlds-circleci-jrs-results"
      automated_run:
        type: boolean
        default: false
      slack_results_channel:
        type: string
        default: ""
      slack_summary_channel:
        type: string
        default: ""
    executor: << parameters.runtime >>
    steps:
      - git_main_repo_checkout
      - mvn_version_info
      - gcp_import_credentials
      - mvn_compile
      - when:
          condition: << parameters.automated_run >>
          steps:
            - jrs_history_retrieve:
                bucket_name: << parameters.summary_bucket >>
                summary_path: << parameters.summary_path >>
      - jrs_regression_execute:
          config_file: << parameters.config_file >>
          regression_path: << parameters.regression_path >>
          slack_results_channel: << parameters.slack_results_channel >>
          slack_summary_channel: << parameters.slack_summary_channel >>
      - when:
          condition: << parameters.automated_run >>
          steps:
            - jrs_history_store:
                bucket_name: << parameters.summary_bucket >>
                summary_path: << parameters.summary_path >>
      - jrs_results_store:
          bucket_name: << parameters.result_bucket >>
          result_path: << parameters.result_path >>
```

#### GCP Storage Rsync Command

```
######################################################################################################################
  # Command Name:                     gcp_storage_rsync
  # Command Version:                  1.0
  #
  # Parameters:
  #       src                         the source local or remote path, mandatory
  #       dest                        the destination local or remote path, mandatory
  #
  #
  # Description:
  #
  #   Copies files or folders (recursively) from the path specified by `src` parameter to the path specified by the
  #   `dest` parameter. The paths specified by both the `src` and `dest` parameters may be local or remote paths.
  #
  #   This command uses a differential rsync and will only transfer new or changed files.
  #
  #   Example Supported Paths:
  #       - /home/myPath
  #       - /home/myPath/file.txt
  #       - gs://my-bucket/folder
  #       - gs://my-bucket/file.txt
  #
  #   NOTE: The build will fail if the `gcloud` command is not available or no CRC32 system library is available.
  #
  # Command Requirements:
  #   - gcloud (>= 323.0.0)
  #
  ######################################################################################################################
  gcp_storage_rsync:
    parameters:
      src:
        type: string
      dest:
        type: string
    steps:
      - run:
          name: "GCP Storage RSync [Source: '<< parameters.src >>', Destination: '<< parameters.dest >>']"
          command: |
            set -x
            gsutil -m rsync -r "<< parameters.src >>" "<< parameters.dest >>"
```

#### GCP Import Credentials Command

```
  ######################################################################################################################
  # Command Name:                     gcp_import_credentials
  # Command Version:                  1.0
  #
  # Parameters:
  #       username_variable           the name of the bash environment variable holding the user name, optional
  #       keyfile_variable            the name of the bash environment variable holding the key file, optional
  #       project_variable            the name of the bash environment variable holding the project name, optional
  #
  #
  # Description:
  #
  #   Imports the gcloud keypair defined in the CircleCI environment variables specified by the
  #   `username_variable`, `keyfile_variable`, and `project_variable` parameters.
  #   The default variable names `GCP_USER_NAME`, `GCP_KEY_FILE`, and `GCP_PROJECT_ID` will be used if not specified.
  #
  #   NOTE: The build will fail if the `gcloud` command is not available or the environment variable is not
  #         properly formatted.
  #
  # Command Requirements:
  #   - gcloud (>= 323.0.0)
  #
  ######################################################################################################################
  gcp_import_credentials:
    parameters:
      username_variable:
        type: string
        default: "GCP_USER_NAME"
      keyfile_variable:
        type: string
        default: "GCP_KEY_FILE"
      project_variable:
        type: string
        default: "GCP_PROJECT_ID"
    steps:
      - run:
          name: Import the Google Cloud Service Account
          command: |
            echo "${<< parameters.keyfile_variable >>}" | base64 -d > "/tmp/gcloud_account_key.json"
            set -x
            gcloud auth activate-service-account "${<< parameters.username_variable >>}" --key-file="/tmp/gcloud_account_key.json" --project="${<< parameters.project_variable >>}"
```

#### JRS History Retrieve

```
  ######################################################################################################################
  # Command Name:                     jrs_history_retrieve
  # Command Version:                  1.0
  #
  # Parameters:
  #       bucket_name                 the name of the GCP bucket containing the JRS summary history files, mandatory
  #       summary_path                the local path of the JRS summary history files, mandatory
  #
  #
  # Description:
  #
  #   Retrieves the JRS Summary History files from prior runs using a combination of CircleCI caching and GCP bucket
  #   storage as specified by the `bucket_name` parameter. The local summary history path specified by the
  #   `summary_path` parameter will be created if it does not exist.
  #
  #   The GCP bucket should be a dedicated storage bucket that must only contain JRS summary history files.
  #
  #   NOTE: The build will fail if the `gcloud` command is not available or no CRC32 system library is available.
  #
  # Command Requirements:
  #   - gcloud (>= 323.0.0)
  #
  ######################################################################################################################
  jrs_history_retrieve:
    parameters:
      bucket_name:
        type: string
      summary_path:
        type: string
    steps:
      - run:
          name: Ensure JRS Summary Folder Exists
          command: |
            set -x
            if [[ ! -d "<< parameters.summary_path >>" ]]; then
              mkdir -p "<< parameters.summary_path >>"
            fi
      - restore_cache:
          name: Restoring JRS Summary History Cache
          keys:
            - v1-jrs-summary-history-{{ epoch }}
            - v1-jrs-summary-history-
      - gcp_storage_rsync:
          src: "gs://<< parameters.bucket_name >>/"
          dest: "<< parameters.summary_path >>"
```

#### JRS History Store

```
  ######################################################################################################################
  # Command Name:                     jrs_history_store
  # Command Version:                  1.0
  #
  # Parameters:
  #       bucket_name                 the name of the GCP bucket containing the JRS summary history files, mandatory
  #       summary_path                the local path of the JRS summary history files, mandatory
  #
  #
  # Description:
  #
  #   Saves the JRS Summary History files from the current run using a combination of CircleCI caching and GCP bucket
  #   storage as specified by the `bucket_name` parameter. The local summary history path specified by the
  #   `summary_path` parameter must exist.
  #
  #   The GCP bucket should be a dedicated storage bucket that must only contain JRS summary history files.
  #
  #   NOTE: The build will fail if the `gcloud` command is not available or no CRC32 system library is available.
  #
  # Command Requirements:
  #   - gcloud (>= 323.0.0)
  #
  ######################################################################################################################
  jrs_history_store:
    parameters:
      bucket_name:
        type: string
      summary_path:
        type: string
    steps:
      - save_cache:
          name: Saving JRS Summary History Cache
          key: v1-jrs-summary-history-{{ epoch }}
          paths:
            - << parameters.summary_path >>
          when: always
      - gcp_storage_rsync:
          src: "<< parameters.summary_path >>"
          dest: "gs://<< parameters.bucket_name >>/"

```

#### JRS Results Store

```
  ######################################################################################################################
  # Command Name:                     jrs_results_store
  # Command Version:                  1.0
  #
  # Parameters:
  #       bucket_name                 the name of the GCP bucket containing the JRS regression results, mandatory
  #       result_path                 the local path of the JRS regression result files, mandatory
  #
  #
  # Description:
  #
  #   Saves the JRS Regression result files from the current run to the GCP bucket storage as specified by the
  #   `bucket_name` parameter. The local result path specified by the `result_path` parameter must exist.
  #
  #   The GCP bucket should be a dedicated storage bucket that must only contain JRS Regression result files.
  #
  #   The final path used to the store the result files in the GCP bucket is as follows:
  #       - gs://${BUCKET_NAME}/${JRS_USER}/${JRS_BRANCH}
  #
  #   The ${JRS_USER} variable will default to a value of `swirlds-automation` if not otherwise provided by CircleCI.
  #   This will be the case with all nightly or other jobs run via the CircleCI cron scheduler. All nightly automated
  #   regression runs should have results stored under this default username.
  #
  #   The ${JRS_BRANCH} variable will be set to the actual branch name provided by the ${CIRCLE_BRANCH} variable, if
  #   available, otherwise it will default to the tag name specified by the ${CIRCLE_TAG} variable, if available. If
  #   neither the ${CIRCLE_BRANCH} or ${CIRCLE_TAG} values are available then then CircleCI job name provided by the
  #   ${CIRCLE_JOB} variable will be used.
  #
  #   NOTE: The build will fail if the `gcloud` command is not available or no CRC32 system library is available.
  #
  # Command Requirements:
  #   - gcloud (>= 323.0.0)
  #
  ######################################################################################################################
  jrs_results_store:
    parameters:
      bucket_name:
        type: string
      result_path:
        type: string
    steps:
      - run:
          name: Saving JRS Regression Results
          command: |
            set -x

            if [[ ! -d "<< parameters.result_path >>" ]]; then exit 13; fi

            if [[ -z "${CIRCLE_USERNAME}" ]]; then
              JRS_USER="swirlds-automation"
            else
              JRS_USER="${CIRCLE_USERNAME}"
            fi

            if [[ -z "${CIRCLE_BRANCH}" ]]; then
              if [[ -n "${CIRCLE_TAG}" ]]; then
                JRS_BRANCH="${CIRCLE_TAG}"
              else
                JRS_BRANCH="${CIRCLE_JOB}"
              fi
            else
              JRS_BRANCH="${CIRCLE_BRANCH}"
            fi

            gsutil -m rsync -r "<< parameters.result_path >>" "gs://<< parameters.bucket_name >>/${JRS_USER}/${JRS_BRANCH}/"
          when: always
```

#### JRS Regression Execute

```
  ######################################################################################################################
  # Command Name:                     jrs_regression_execute
  # Command Version:                  1.0
  #
  # Parameters:
  #       config_file                 the relative path to the JRS regression config, mandatory
  #       regression_path             the local path to the regression folder, optional (default: "regression")
  #       slack_results_channel       the slack test results channel override, optional (default: "")
  #       slack_summary_channel       the slack summary channel override, optional (default: "")
  #
  #
  # Description:
  #
  #   Executes the JRS regression run using the provided `config_file` parameter. The `config_file` parameter is treated
  #   as a path relative to the `regression_path` parameter. The `regression_path` parameter may be either a path
  #   relative to the current working directory or an absolute path.
  #
  #   The optional `slack_results_channel` and `slack_summary_channel` parameters allow using an alternate slack channel
  #   for the respective notification types. If one or both of these are provided, then the related setting in the JSON
  #   files will be ignored.
  #
  #   This command depends on the following CircleCI Context environment variables:
  #     - JRS_SSH_USER_NAME:  the username associated with the `JRS_SSH_KEY_FILE` used by JRS to connect to the remote instances
  #     - JRS_SSH_KEY_FILE:   the base64 encoded private SSH key used by JRS to connect to remote instances
  #     - JRS_WEB_HOSTNAME:   the IP address or hostname of the JRS web server
  #     - JRS_WEB_PORT:       the port number on which the JRS web server is listening
  #
  #
  #   NOTE: The build will fail if the `gcloud` command is not available.
  #
  # Command Requirements:
  #   - gcloud (>= 323.0.0)
  #   - openjdk (>= 12.0.2)
  #
  ######################################################################################################################
  jrs_regression_execute:
    parameters:
      config_file:
        type: string
      regression_path:
        type: string
        default: "regression"
      slack_results_channel:
        type: string
        default: ""
      slack_summary_channel:
        type: string
        default: ""
    steps:
      - run:
          name: Configure JRS Regression Keys
          command: |
            set -x
            echo "${JRS_SSH_KEY_FILE}" | base64 -d > "/tmp/jrs-ssh-keyfile"
            cp "/tmp/jrs-ssh-keyfile" "/tmp/jrs-ssh-keyfile.pem"
            chmod 0600 /tmp/jrs-ssh-keyfile*
            ssh-keygen -p -N "" -m pem -f "/tmp/jrs-ssh-keyfile.pem"
            ssh-keygen -y -f "/tmp/jrs-ssh-keyfile" > "/tmp/jrs-ssh-keyfile.pub"
      - run:
          name: "Execute JRS Regression (<< parameters.config_file >>)"
          command: |
            set -x

            REGRESSION_PATH="<< parameters.regression_path >>"
            if [[ ! -d "${REGRESSION_PATH}" ]]; then exit 15; fi

            if [[ -z "${CIRCLE_USERNAME}" ]]; then
              JRS_USER="swirlds-automation"
            else
              JRS_USER="${CIRCLE_USERNAME}"
            fi

            if [[ -z "${CIRCLE_BRANCH}" ]]; then
              if [[ -n "${CIRCLE_TAG}" ]]; then
                JRS_BRANCH="${CIRCLE_TAG}"
              else
                JRS_BRANCH="${CIRCLE_JOB}"
              fi
            else
              JRS_BRANCH="${CIRCLE_BRANCH}"
            fi

            JRS_OPTIONS=""
            SLACK_SUMMARY="<< parameters.slack_summary_channel >>"
            SLACK_RESULTS="<< parameters.slack_results_channel >>"

            if [[ -n "${SLACK_SUMMARY}" ]]; then
              JRS_OPTIONS="${JRS_OPTIONS} -Djrs.circleci.slack.summary=${SLACK_SUMMARY}"
            fi

            if [[ -n "${SLACK_RESULTS}" ]]; then
              JRS_OPTIONS="${JRS_OPTIONS} -Djrs.circleci.slack.results=${SLACK_RESULTS}"
            fi


            pushd "${REGRESSION_PATH}" > /dev/null 2>&1
              CONFIG_PATH="<< parameters.config_file >>"
              if [[ ! -f "${CONFIG_PATH}" ]]; then
                echo
                echo "Configuration File '${CONFIG_PATH}' does not exist......"
                echo
                exit 20
              fi

              if [[ -z "${JAVA_OPTS}" ]]; then
                JAVA_OPTS="-Xmx4g"
              fi

              java ${JAVA_OPTS} \
              -Djrs.circleci=true \
              -Djrs.circleci.user="${JRS_USER}" \
              -Djrs.circleci.branch="${JRS_BRANCH}" \
              -Djrs.circleci.ssh.login="${JRS_SSH_USER_NAME}" \
              -Djrs.circleci.ssh.keyfile="/tmp/jrs-ssh-keyfile" \
              -Djrs.circleci.web.hostname="${JRS_WEB_HOSTNAME}" \
              -Djrs.circleci.web.port="${JRS_WEB_PORT}" \
              ${JRS_OPTIONS} \
              -Dlog4j.configurationFile=log4j2-jrs.xml \
              -Dspring.output.ansi.enabled=ALWAYS \
              -jar regression.jar "${CONFIG_PATH}"
            popd > /dev/null 2>&1
          no_output_timeout: 30m
```

### PR DESCRIPTION
**Pull request by ljianghedera**
**Related issue(s)**:
Closes #995

**Summary of the change**:
1. Enables regression test summary history
2. Use `jrs-regression` job for all workflows and remove all unused jobs
3. Stores the test results in GCP in [Test Results ](https://console.cloud.google.com/storage/browser/hedera-services-jrs-test-results/hedera-services-automatio?&project=hedera-regression)
4. The link to GCP test results to be shown in **Results Folder** in slack is updated in platform -regression [PR](https://github.com/swirlds/swirlds-platform-regression/pull/840) 

 **Tests with Summary links and Results Folder link:** 
Failures of the test are similar to failures on master, they should have separate issues. This issue addresses only for summary history.

1. Migration : https://hedera-hashgraph.slack.com/archives/C018CBQBTGB/p1612196329112800
2. NetError: https://hedera-hashgraph.slack.com/archives/C018CBQBTGB/p1612198716114800
3. Restart : https://hedera-hashgraph.slack.com/archives/C018CBQBTGB/p1612198876115400
4. Reconnect : https://hedera-hashgraph.slack.com/archives/C018CBQBTGB/p1612201473118500
5. HTS Perf test : https://hedera-hashgraph.slack.com/archives/C018CBQBTGB/p1612204269121600

### CODE REVIEW

**Review by qnswirlds — COMMENTED** [review]
Overall looks good to me. But I think it's better to extract all of those scripts and save them under `.circleci/scripts`. `.circleci/config.yml` should be simplified as much as we can.

File: .circleci/config.yml [hunk 1]
```diff
@@ -35,6 +35,464 @@ commands:
...
+  #
+  #   Saves the JRS Regression result files from the current run to the GCP bucket storage as specified by the
+  #   `bucket_name` parameter. The local result path specified by the `result_path` parameter must exist.
+  #
+  #   The GCP bucket should be a dedicated storage bucket that must only contain JRS Regression result files.
+  #
+  #   The final path used to the store the result files in the GCP bucket is as follows:
+  #       - gs://${BUCKET_NAME}/${JRS_USER}/${JRS_BRANCH}
+  #
+  #   The ${JRS_USER} variable will default to a value of `swirlds-automation` if not otherwise provided by CircleCI.
```
Comment (ljianghedera) [inline .circleci/config.yml:240]: Better change `swirlds-automation` to `hedera-services-automation`.

**Comment by ljianghedera** [comment]
> Overall looks good to me. But I think it's better to extract all of those scripts and save them under `.circleci/scripts`. `.circleci/config.yml` should be simplified as much as we can.

Technically that would be the right action. But here, the next move is to extract all the commands and jobs that can be shared with platform and other projects into CircleCi's orbs once getting the license issue resolved. So it would be fine to leave these scripts as it is for now.

**Comment by nathanklick** [comment]
> Overall looks good to me. But I think it's better to extract all of those scripts and save them under `.circleci/scripts`. `.circleci/config.yml` should be simplified as much as we can.

Actually, the best practices as recommended by CircleCI is to use inline bash and not to depend on external bash scripts whenever possible.  

Once we get the licensing resolved as @ljianghedera stated we can move to orbs which will improve maintainability, but ultimately still requires the bash code to be inline..

File: .circleci/config.yml (hunk 1 above)
Comment (Neeharika-Sompalli) [reply .circleci/config.yml]: Fixed. Thanks @ljianghedera

**Comment by Neeharika-Sompalli** [comment]
> > Overall looks good to me. But I think it's better to extract all of those scripts and save them under `.circleci/scripts`. `.circleci/config.yml` should be simplified as much as we can.
> 
> Actually, the best practices as recommended by CircleCI is to use inline bash and not to depend on external bash scripts whenever possible.
> 
> Once we get the licensing resolved as @ljianghedera stated we can move to orbs which will improve maintainability, but ultimately still requires the bash code to be inline. The only difference is that the Orb contains the bash and the inlining is handled by the orb tools.

Thanks @nathanklick @ljianghedera. @qnswirlds Based on Nathan's comments, I can create an issue to move to Orbs once licensing is resolved.

**Review by qnswirlds — APPROVED** [review]

**Pull request merged**